xonsh/cli_utils.py

```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+        and HAS_PYGMENTS
+    ):
+        return RstHelpFormatter
+    return ap.RawTextHelpFormatter
+
+
+class ArgParser(ap.ArgumentParser):
+    """Sub-class of ArgumentParser with special methods to nest commands"""
+
+    def __init__(self, **kwargs):
+        if "formatter_class" not in kwargs:
+            kwargs["formatter_class"] = get_argparse_formatter_class()
+
+        super().__init__(**kwargs)
+        self.commands = None
+
+    def add_command(
+        self, func: tp.Callable, args: tp.Optional[tp.Iterable[str]] = None, **kwargs
+    ):
+        """
+            create a sub-parser and call this function during dispatch
+
+        Parameters
+        ----------
+        func
+            a type-annotated function that will be used to create ArgumentParser instance.
+            All parameters that start with ``_`` will not be added to parser arguments.
+            Use _stdout, _stack ... to receive them from callable-alias/commands.
+            Use _parser to get the generated parser instance.
+            Use _args to get what is passed from sys.argv
+            Use _parsed to get result of ``parser.parse_args``
+        args
+            if given only add these arguments to the parser.
+            Otherwise all parameters to the function without `_` prefixed
+            in their name gets added to the parser.
+        kwargs
+            passed to ``subparser.add_parser`` call
+
+        Returns
+        -------
+            result from ``subparser.add_parser``
+        """
+        if not self.commands:
+            self.commands = self.add_subparsers(title="commands", dest="command")
+
+        doc = get_doc(func)
+        kwargs.setdefault("description", doc)
+        kwargs.setdefault("help", doc)
+        parser = self.commands.add_parser(kwargs.pop("prog", func.__name__), **kwargs)
+        add_args(parser, func, allowed_params=args)
         return parser
 
 
-def dispatch(**ns):
+def dispatch(parser: ap.ArgumentParser, args=None, **ns):
     """call the sub-command selected by user"""
     import inspect
 
+    parsed = parser.parse_args(args)
+    ns["_parsed"] = parsed
+    ns.update(vars(parsed))
+
     func = ns[_FUNC_NAME]
     sign = inspect.signature(func)
     kwargs = {}
-    for name, _ in sign.parameters.items():
-        kwargs[name] = ns[name]
+    for name, param in sign.parameters.items():
+        default = None
+        if param.default != inspect.Parameter.empty:
+            default = param.default
+        kwargs[name] = ns.get(name, default)
     return func(**kwargs)
+
+
+class ArgParserAlias:
+    """Provides a structure to the Alias. The parser is lazily loaded.
+
+    can help create ``argparse.ArgumentParser`` parser from function
+    signature and dispatch the functions.
+
+    Examples
+    ---------
+        For usage please check ``xonsh.completers.completer.py`` module.
+    """
+
+    def __init__(self, threadable=True, **kwargs) -> None:
```

Comment:
How should this be used with simple one-function aliases?
I reckon a simple `@arg_parser_alias(threadable=...)` decorator can be useful (e.g. the `xoreutils` aliases)

Reply:
![image](https://user-images.githubusercontent.com/6702219/126885236-6849beff-dce0-4757-8f5d-3b3c18e3a3e3.png)

this is how single functions can be added.